src/main/java/net/minecraftforge/fml/common/ModTrackingList.java

```diff
@@ -0,0 +1,316 @@
+        }
+        return changed;
+    }
+
+    @Override
+    public void add(int index, @Nonnull T element)
+    {
+        delegate.add(index, element);
+        trackModContainer(element);
+    }
+
+    @Override
+    public boolean addAll(int index, @Nonnull Collection<? extends T> elements)
+    {
+        boolean changed = delegate.addAll(index, elements);
+        if (changed)
+        {
+            ModContainer modContainer = Loader.instance().activeModContainer();
+            if (modContainer != null)
+            {
+                for (T element : elements)
+                {
+                    modContainerMap.put(element, modContainer);
+                }
+            }
+        }
+        return changed;
+    }
+
+    @Override
+    public boolean addAll(@Nonnull Collection<? extends T> collection)
+    {
+        boolean changed = delegate.addAll(collection);
+        if (changed)
+        {
+            ModContainer modContainer = Loader.instance().activeModContainer();
+            if (modContainer != null)
+            {
+                for (T element : collection)
+                {
+                    modContainerMap.put(element, modContainer);
+                }
+            }
+        }
+        return changed;
+    }
+
+    @Override
+    public T remove(int index)
+    {
+        T removed = delegate.remove(index);
+        if (removed != null)
+        {
+            modContainerMap.remove(removed);
+        }
+        return removed;
+    }
+
+    @Override
+    public boolean remove(@Nonnull Object object)
+    {
+        boolean changed = delegate.remove(object);
+        if (changed)
+        {
+            modContainerMap.remove(object);
+        }
+        return changed;
+    }
+
+    @Override
+    public boolean removeAll(@Nonnull Collection<?> collection)
+    {
+        boolean changed = delegate.removeAll(collection);
+        if (changed)
+        {
+            modContainerMap.keySet().removeAll(collection);
+        }
+        return changed;
+    }
+
+    @Override
+    public T set(int index, @Nonnull T element)
+    {
+        T previous = delegate.set(index, element);
+        if (previous != null)
+        {
+            modContainerMap.remove(previous);
+        }
+        trackModContainer(element);
+        return previous;
+    }
+
+    @Nonnull
+    @Override
+    public List<T> subList(int fromIndex, int toIndex)
+    {
+        List<T> delegateSubList = delegate.subList(fromIndex, toIndex);
+        return new ModTrackingList<T>(delegateSubList, modContainerMap);
+    }
+
+    @Override
+    public boolean retainAll(@Nonnull Collection<?> collection)
+    {
+        boolean changed = delegate.retainAll(collection);
+        if (changed)
+        {
+            modContainerMap.keySet().retainAll(collection);
+        }
+        return changed;
+    }
+
+    @Nonnull
+    @Override
+    public Iterator<T> iterator()
+    {
+        return new ModTrackedIterator(delegate.iterator());
```

Comment:
One is an `Iterator` and one is a `ListIterator`, I'm not sure it's correct to use `ListIterator` for both. There may be some extra performance in using `Iterator`?